server/dataset_context.h

```diff
@@ -78,12 +79,47 @@ struct SqlExpressionDatasetScope: public SqlExpressionMldbScope {
 
     struct RowScope: public SqlRowScope {
         RowScope(const MatrixNamedRow & row,
-                   const BoundParameters * params = nullptr)
-            : row(row), params(params)
+                 const BoundParameters * params = nullptr)
+            : row(&row), rowName(nullptr), expr(nullptr), params(params)
         {
         }
 
-        const MatrixNamedRow & row;
+        RowScope(const RowName & rowName,
+                 const ExpressionValue & row,
+                 const BoundParameters * params = nullptr)
+            : row(nullptr), rowName(&rowName), expr(&row), params(params)
```

Comment:
I think "row", as function parameter, being passed to "expr" is confusing, especially when "row" as member var is set to something else.